### ISSUE TITLE
co19/LanguageFeatures/nnbd/flow-analysis/reachability_* fail

### ISSUE THREAD

**Issue by scheglov** [dart-lang/sdk#42193]
For example `co19/LanguageFeatures/nnbd/flow-analysis/reachability_A05_t03`.

```dart
/**
 * @assertion other literal: If N is some other literal than the above, then:
 *  Let null(N) = unreachable(before(N)).
 *  Let notNull(N) = before(N)
 * @description Checks reachability after int literal
 *
 * @author sgrekhov@unipro.ru
 * @issue 41985
 */
// SharedOptions=--enable-experiment=non-nullable
// Requirements=nnbd-strong

main() {
  late int i;
  if (0 == null) {
    i = 42;  // Variable is initialized in a dead code. This leaves it definitely unassigned
  }
  i;  // It is an error to read a local late variable when it is definitely unassigned.
//^
// [analyzer] unspecified
// [cfe] unspecified
}
```

The expectation seems reasonable to me.
`0` is never equal to `null`, so the assignment is never reachable.

We currently don't implement `operator==` fully as specified, we only have promotion of variables to non-nullable. Specifically, we don't have reachability updated.

But when I try to evaluate what the specification says, I don't get the unreachable state.

- **operator==** If `N` is an expression of the form `E1 == E2` then:
  - Let `before(E1) = before(N)`
  - Let `before(E2) = after(E1)`
  - Let `true(N) = join(join(null(E1), null(E2)), join(notNull(E1), notNull(E2)))`
  - Let `false(N) = join(join(null(E1), notNull(E2)),
                         join(notNull(E1), null(E2)),
                         join(notNull(E1), notNull(E2)))`

We need to evaluate `true(0 == null)`.
We have `E1 = 0`, `E2 = null`; `null(E1) = unreachable`, `null(E2) = reachable`, `nonNull(E1) = reachable`, `nonNull(E2) = unreachable`. But `join(reachable, unreachable) = reachable`. So, both `join(null(E1), null(E2)) = reachable`, and `join(notNull(E1), notNull(E2)) = reachable`, and so the whole `true(0 == null) = reachable`.

So, the assignment is reachable, and the error for the `late int i` is not reported.
Is there a mistake in my calculations?

@leafpetersen @stereotype441 You noticed another issue in [here](https://github.com/dart-lang/language/pull/763/files#r364003138). Do you want to fix the specification? Or do we keep it as a known issue, and keep the implementation as is? What should we do with `co19` tests (there are 26 related tests)?

**Comment by stereotype441** [comment]
Leaf and I talked about this a long time ago.  See https://github.com/dart-lang/language/pull/763#discussion_r366642882

My two cents: I think we're getting close enough to the NNBD deadline that for now we should adjust the spec to match what we currently have implemented; if time allows, we can always follow up with an improvement later.

**Comment by leafpetersen** [comment]
I think this is not planned, is that correct @stereotype441 ?

**Comment by stereotype441** [comment]
I suspect this was addressed by e403e6ae77c1c6fabd5dd5a9db26b63f5a416057, but unfortunately I don't have time to confirm right now.  I'll double check when I return from vacation on Wednesday.

**Comment by leafpetersen** [comment]
This seems to have been fixed.

**Comment by stereotype441** [comment]
Confirmed that this was fixed by e403e6ae77c1c6fabd5dd5a9db26b63f5a416057.
